### PR TITLE
DNN: add ONNX where node support

### PR DESCRIPTION
**Pull request by zihaomu**
Merge with test data: https://github.com/opencv/opencv_extra/pull/1054

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/layers/nary_eltwise_layers.cpp [hunk 1]
```diff
@@ -499,6 +502,120 @@ class NaryEltwiseLayerImpl CV_FINAL : public NaryEltwiseLayer
                 f, scale, ninputs, max_ndims, shapes[0], inp, out, (const size_t **) steps, ptrs);
     }
 
+    template <typename T, typename Functor>
+    void trinary_forward(const Functor& f, const std::vector<Mat>& inputs, std::vector<Mat>& outputs)
```
Comment (rogday) [inline modules/dnn/src/layers/nary_eltwise_layers.cpp:506]: To be honest it looks like we could change the original function to accept multiple input vector.
E.g. make it accept (data + step) buffer. And everywhere where there is $i suffix (e.g. ptr1_, etc), we could use buffer as well.
I think we had performance tests somewhere, we should check that it didn't degrade. If it does, we could have 1) fast binary path and 2) slower naray path.

Comment (zihaomu) [reply modules/dnn/src/layers/nary_eltwise_layers.cpp:506]: Hi, @rogday. Thanks for the code review.
I'm not very familiar with this part of the code, can you give some examples in more detail so I can complete it?

**Comment by asmorkalov** [comment]
@dkurt Could you, please, review the PR too?

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -2492,6 +2492,11 @@ TEST_P(Test_ONNX_layers, OpenAI_CLIP_head)
     testONNXModels("clip-vit-base-head");
 }
 
+TEST_P(Test_ONNX_layers, where_node)
+{
+    testONNXModels("where_layer");
+}
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2498]: Please add a test for broadcasting scenario.

Comment (zihaomu) [reply modules/dnn/test/test_onnx_importer.cpp:2498]: image

Hi @dkurt. Thanks for your code review, and I have updated the test case.

**Review by dkurt — APPROVED** [review]

**Comment by asmorkalov** [comment]
@zihaomu `where` node is mentioned in some community reported issues. Could you update them and at lest add reverence to the PR.

**Comment by zihaomu** [comment]
Link to https://github.com/opencv/opencv/issues/23470#issuecomment-1504803680

**Pull request merged**